### ISSUE TITLE
[BUG] ModelCheckpoint callback wont work with log on_epoch = False

### ISSUE THREAD

**Issue by fmellomascarenhas** [PyTorchLightning/pytorch-lightning#10718]
I don't think that the problem below is necessarily a bug, but maybe the error message can be improved for future users.

## 🐛 Bug

If `on_epoch = False`
```
def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid/valid_loss", loss, on_step = True, on_epoch = False)
```

The validation_metric is not available for modelcheckpoint callback, throwing the following error:
`MisconfigurationException: ModelCheckpoint(monitor='valid/valid_loss') not found in the returned metrics: ['train_loss']. HINT: Did you call self.log('valid/valid_loss', value) in the LightningModule?`

If `on_epoch = True`, then wandb logger creates a mess in the logging process, saving a metric for epoch and a metric for steps:
![image](https://user-images.githubusercontent.com/23004953/143109626-d7912278-a9c1-4f7a-b482-3dc9ffae2111.png)

In my mind, because i was running validation within the epoch, and not only the end, I had to log on_steps = True and on_epoch = False. Apparently I can remove both parameters from the log and it works fine. So maybe the error message of the modelcheckpoint can be improved, since I spent 2 days trying to figure out why I couldnt save my model using this callback. 

### To Reproduce
https://colab.research.google.com/drive/1CwznZUFC3ctKT4bQbpLe-p4JOF9o_9VP?usp=sharing

Thanks!

cc @carmocca @edward-io @ananthsub

**Comment by rohitgr7** [comment]
the reason might be:
```
self.log("valid/valid_loss", loss, on_step = True, on_epoch = False)
```
`valid/valid_loss` is a step level metric and checkpointing happens on epoch level, so checkpointing based on the metric that is not accumulated at all won't be correct?

**Comment by fmellomascarenhas** [comment]
@rohitgr7 regarding the statement `"checkpointing happens on epoch level"`, this is not necessarily true, right? Checkpoing callback has the argument `every_n_train_steps`. My epoch takes a very long time, so I need to run checkpointing within the epoch, but as an user, it was not clear to me how to manage the logging/checkpoint/trainer so they all can run validation within epoch. If you check my trainer in the boring_model, you will see that I am using `"val_check_interval = 0.1"`.

**Comment by rohitgr7** [comment]
yeah for that case, it should be there.
but still, I don't think it is the right way to do checkpointing: monitoring just one of the validation batch metric to decide whether to checkpoint or not.

can you make your colab notebook public?

**Comment by tchaton** [comment]
Dear @fmellomascarenhas,

Yes, I think this isn't supported as it is technically not the right way to perform checkpointing as @rohitgr7 shared.
Possibly we could have a better error message. 

Best,
T.C

**Comment by fmellomascarenhas** [comment]
> can you make your colab notebook public?
It should be now

> I don't think it is the right way to do checkpointing: monitoring just one of the validation batch metric to decide
I agree, but how does checkpointing works if `every_n_train_steps = n_steps`? Maybe for checkpointing within epoch, the mean of the last N validation results can be used.

@tchaton @rohitgr7

**Comment by rohitgr7** [comment]
using `every_n_train_steps` with a monitor is tricky because we can't guarantee a proper monitor value to compare if all of them are logged on batch level. I have seen users using this to save checkpoints at regular training steps intervals without any monitor. If you log your validation metric with `on_epoch=True`, it should work because then we have a metric that is worth considering as a monitor value to compare against.

> Maybe for checkpointing within epoch, the mean of the last N validation results available can be used.

Can you elaborate more on this? N here refers to N training steps (optimization steps).

**Comment by fmellomascarenhas** [comment]
I guess there were some misinterpretations on my end about how to perform validation within the training steps. Gonna close with now, but maybe the error message can still have some hint for the user in this particular error. Thanks for the replies!
